tests/zfs-tests/tests/functional/mount/umount_unlinked_drain.ksh

```diff
@@ -0,0 +1,119 @@
+#!/bin/ksh -p
+
+#
+# This file and its contents are supplied under the terms of the
+# Common Development and Distribution License ("CDDL"), version 1.0.
+# You may only use this file in accordance with the terms of version
+# 1.0 of the CDDL.
+#
+# A full copy of the text of the CDDL should have accompanied this
+# source.  A copy of the CDDL is also available via the Internet at
+# http://www.illumos.org/license/CDDL.
+#
+
+#
+# Copyright 2018 Datto Inc.
+#
+
+. $STF_SUITE/include/libtest.shlib
+
+#
+# DESCRIPTION:
+# Test async unlinked drain to ensure mounting is not held up when there are
+# entries in the unlinked set. We also try to test that the list is able to be
+# filled up and drained at the same time.
+#
+# STRATEGY:
+# 1. Use zfs_unlink_suspend_progress tunable to disable freeing to build up
+#    the unlinked set
+# 2. Make sure mount happens even when there are entries in the unlinked set
+# 3. Drain and build up the unlinked list at the same time to test for races
+#
+
+function cleanup
+{
+	log_must set_tunable32 zfs_unlink_suspend_progress $default_unlink_sp
+	for fs in $(seq 1 3); do
+		mounted $TESTDIR.$fs || zfs mount $TESTPOOL/$TESTFS.$fs
+		rm -f $TESTDIR.$fs/file-*
+		zfs set xattr=on $TESTPOOL/$TESTFS.$fs
+	done
+}
+
+function unlinked_size_is
+{
+	MAX_ITERS=5 # iteration to do before we consider reported number stable
```

Comment:
Isn't `iput` just `VN_RELE` and what we usually use ? Could we have that instead? Or, since ZOL is taking over, should I add `iput` and `ZTOI` macros?

Reply:
I originally had a `sleep 10` here but I belive this version is actually more robust than having a static sleep or TXGs threshold.

This code provides us with dynamic behavior since we keep looping until the unlinked set size has not changed for the last 5 loop iterations. That allows us to avoid unecessary test delays while still trying to ensure that the unlink set processing has "finished".

Having said that, it's easy for me to remove the loop and put back the `sleep 10` if you feel that is more appropriate.

Reply:
Adding a brief delay between iterations (say 0.1) might be useful, my only concern would be that if the checks are done to quickly we might be mistaken about it stabilizing.

Reply:
From looking at the code, I couldn't figure out why we would ever do more than one iteration.  Why wouldn't the unlinked size be the expected size right away?

Reply:
Since the test does the file deletions in a forked process my concern was that we may not have processed all of the backgrounded file deletions by the time we go to check the unlinked set size.
So perhaps I just put a `sleep 10` after the `rm` calls instead of doing this function loop

Reply:
I'd like to avoid adding these kinds of gratuitous sleeps, the ZTS runs long enough already and there's still no assurance any time you pick with be long enough.  The existing approach seems more robust.

Reply:
Ah, I didn't see that you were running `rm` in the background.  How does that interact with `log_must`?  Can it actually log the output and check the return code after it's in the background?

Could we avoid the loop/sleep by waiting for the background `rm`'s to complete?  Normally `wait` would do that, but I don't know what the interaction is with `log_must eval`.

Reply:
Alternatively, do we really need to do the 100 `rm`'s in the background?  I would guess that the fork/exec/startup overhead is much more than the actual `unlink` syscall.  You could probably improve performance by doing `log_must rm $TESTDIR.$fs/file-*` in the foreground.

Reply:
I could have done `log_must rm $TESTDIR.$fs/file-*` sequentially in the foreground but I thought it was a better test to do it in "parallel".